File: web/src/lib/format.ts
const DAY = 86400_000;

function startOfDay(ts: number): number {
  const d = new Date(ts);
  d.setHours(0, 0, 0, 0);
  return d.getTime();
}

export function formatTime(ts: number): string {
  return new Date(ts).toLocaleTimeString(undefined, { hour: '2-digit', minute: '2-digit' });
}

/** Chat-list style: time today, "Yesterday", weekday this week, else date. */
export function formatListTime(ts: number | null, now = Date.now()): string {
  if (!ts) return '';
  const today = startOfDay(now);
  if (ts >= today) return formatTime(ts);
  if (ts >= today - DAY) return 'Yesterday';
  if (ts >= today - 6 * DAY) return new Date(ts).toLocaleDateString(undefined, { weekday: 'long' });
  return new Date(ts).toLocaleDateString();
}

/** Date separator label inside a chat. */
export function formatDayLabel(ts: number, now = Date.now()): string {
  const today = startOfDay(now);
  if (ts >= today) return 'Today';
  if (ts >= today - DAY) return 'Yesterday';
  if (ts >= today - 6 * DAY) return new Date(ts).toLocaleDateString(undefined, { weekday: 'long' });
  return new Date(ts).toLocaleDateString(undefined, { day: 'numeric', month: 'long', year: 'numeric' });
}

export function formatDateTime(ts: number): string {
  return new Date(ts).toLocaleString(undefined, { dateStyle: 'medium', timeStyle: 'short' });
}

export function dayKey(ts: number): string {
  const d = new Date(ts);
  return `${d.getFullYear()}-${d.getMonth()}-${d.getDate()}`;
}

export function formatBytes(n: number | null | undefined): string {
  if (n === null || n === undefined) return '';
  if (n < 1024) return `${n} B`;
  if (n < 1024 * 1024) return `${(n / 1024).toFixed(0)} KB`;
  if (n < 1024 * 1024 * 1024) return `${(n / 1024 / 1024).toFixed(1)} MB`;
  return `${(n / 1024 / 1024 / 1024).toFixed(2)} GB`;
}

export function formatDuration(sec: number | null | undefined): string {
  if (!sec && sec !== 0) return '';
  const s = Math.max(0, Math.round(sec));
  const m = Math.floor(s / 60);
  const h = Math.floor(m / 60);
  const ss = String(s % 60).padStart(2, '0');
  return h ? `${h}:${String(m % 60).padStart(2, '0')}:${ss}` : `${m}:${ss}`;
}

export function initials(name: string): string {
  const clean = name.replace(/^[~+]/, '').trim();
  if (!clean) return '?';
  if (/^\+?\d/.test(clean)) return '#';
  const parts = clean.split(/\s+/).filter(Boolean);
  const first = [...(parts[0] ?? '')][0] ?? '';
  const second = parts.length > 1 ? ([...(parts[parts.length - 1] ?? '')][0] ?? '') : '';
  return (first + second).toUpperCase();
}

const SENDER_COLORS = ['#e17076', '#7bc862', '#65aadd', '#a695e7', '#ee7aae', '#6ec9cb', '#faa774', '#53bdeb', '#ffb74d', '#4db6ac', '#ba68c8', '#f06292'];

export function colorFor(id: string): string {
  let h = 0;
  for (let i = 0; i < id.length; i++) h = (h * 31 + id.charCodeAt(i)) | 0;
  return SENDER_COLORS[Math.abs(h) % SENDER_COLORS.length]!;
}

File: android_client/core/src/main/kotlin/io/github/revocx35/walogger/core/Format.kt
package io.github.revocx35.walogger.core

import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle
import java.util.Locale
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.roundToLong

/** Display formatting, ported from `web/src/lib/format.ts`. */
class Formatter(
    private val locale: Locale = Locale.getDefault(),
    private val zone: ZoneId = ZoneId.systemDefault(),
    use24h: Boolean = true,
) {
    private val time = DateTimeFormatter.ofPattern(if (use24h) "HH:mm" else "hh:mm a", locale)
    private val weekday = DateTimeFormatter.ofPattern("EEEE", locale)
    private val shortDate = DateTimeFormatter.ofLocalizedDate(FormatStyle.SHORT).withLocale(locale)
    private val longDate = DateTimeFormatter.ofLocalizedDate(FormatStyle.LONG).withLocale(locale)
    private val mediumDate = DateTimeFormatter.ofLocalizedDate(FormatStyle.MEDIUM).withLocale(locale)

    private fun date(ts: Long): LocalDate = Instant.ofEpochMilli(ts).atZone(zone).toLocalDate()

    fun time(ts: Long): String = time.format(Instant.ofEpochMilli(ts).atZone(zone))

    /** Chat-list style: time today, "Yesterday", weekday this week, else date. */
    fun listTime(ts: Long?, now: Long = System.currentTimeMillis()): String {
        if (ts == null || ts == 0L) return ""
        val today = date(now)
        val d = date(ts)
        return when {
            !d.isBefore(today) -> time(ts)
            d == today.minusDays(1) -> "Yesterday"
            !d.isBefore(today.minusDays(6)) -> weekday.format(d)
            else -> shortDate.format(d)
        }
    }

    /** Date separator label inside a chat. */
    fun dayLabel(ts: Long, now: Long = System.currentTimeMillis()): String {
        val today = date(now)
        val d = date(ts)
        return when {
            !d.isBefore(today) -> "Today"
            d == today.minusDays(1) -> "Yesterday"
            !d.isBefore(today.minusDays(6)) -> weekday.format(d)
            else -> longDate.format(d)
        }
    }

    fun dateTime(ts: Long): String = "${mediumDate.format(date(ts))}, ${time(ts)}"

    fun dayKey(ts: Long): Long = date(ts).toEpochDay()
}

fun formatBytes(n: Long?): String {
    if (n == null) return ""
    return when {
        n < 1024 -> "$n B"
        n < 1024 * 1024 -> "${(n / 1024.0).roundToLong()} KB"
        n < 1024L * 1024 * 1024 -> String.format(Locale.US, "%.1f MB", n / 1024.0 / 1024.0)
        else -> String.format(Locale.US, "%.2f GB", n / 1024.0 / 1024.0 / 1024.0)
    }
}

fun formatDuration(sec: Double?): String {
    if (sec == null) return ""
    val s = max(0L, sec.roundToLong())
    val m = s / 60
    val h = m / 60
    val ss = (s % 60).toString().padStart(2, '0')
    return if (h > 0) "$h:${(m % 60).toString().padStart(2, '0')}:$ss" else "$m:$ss"
}

private val LEADING_MARK = Regex("^[~+]")
private val PHONE_LIKE = Regex("^\\+?[0-9]")
private val WS = Regex("\\s+")

private fun firstCodePoint(s: String): String = if (s.isEmpty()) "" else String(Character.toChars(s.codePointAt(0)))

/** Avatar initials: "#" for phone numbers, first letters of the first and last word otherwise. */
fun initials(name: String): String {
    val clean = name.replace(LEADING_MARK, "").trim()
    if (clean.isEmpty()) return "?"
    if (PHONE_LIKE.containsMatchIn(clean)) return "#"
    val parts = clean.split(WS).filter { it.isNotEmpty() }
    val first = firstCodePoint(parts.firstOrNull() ?: "")
    val second = if (parts.size > 1) firstCodePoint(parts.last()) else ""
    return (first + second).uppercase()
}

/** Same palette and hash as the web UI, so a contact has the same color everywhere. */
val SENDER_COLORS: LongArray = longArrayOf(
    0xFFE17076, 0xFF7BC862, 0xFF65AADD, 0xFFA695E7, 0xFFEE7AAE, 0xFF6EC9CB,
    0xFFFAA774, 0xFF53BDEB, 0xFFFFB74D, 0xFF4DB6AC, 0xFFBA68C8, 0xFFF06292,
)

/** ARGB color for an id (identical to the web's `colorFor`, including 32-bit overflow). */
fun colorFor(id: String): Long {
    var h = 0
    for (ch in id) h = h * 31 + ch.code
    return SENDER_COLORS[(abs(h.toLong()) % SENDER_COLORS.size).toInt()]
}
